### PR TITLE
controllers: abstract Kubernetes queue into common package

### PR DESCRIPTION
**Pull request by howardjohn**
We have made a lot of new controllers use the Kubernetes queue. Generally this is a good thing, but the API is way too low level. This leads to complex logic throughout much of our code, and its all slightly different with slightly different bugs.

This abstracts this into a new `controllers.Queue` type. In addition, secretcontroller, webhook, gateway deployment, and default revision controllers are moved over to the new queue

### CODE REVIEW

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 1]
```diff
@@ -79,11 +78,25 @@ type patcher func(gvr schema.GroupVersionResource, name string, namespace string
...
+			c := client.Dynamic().Resource(gvr).Namespace(namespace)
+			t := true
+			_, err := c.Patch(context.Background(), name, types.ApplyPatchType, data, metav1.PatchOptions{
+				Force:        &t,
+				FieldManager: ControllerName,
+			}, subresources...)
+			return err
+		},
+	}
+	dc.queue = controllers.NewQueue(controllers.WithReconciler(dc.Reconcile))
```
Comment (stevenctl) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:94]: nit: can be part of the previous statement

File: pkg/kube/controllers/queue.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+	return func(q *Queue) {
+		q.work = func(key interface{}) error {
+			return f(key.(types.NamespacedName))
+		}
+	}
+}
+
+// NewQueue creates a new queue
+func NewQueue(options ...func(*Queue)) Queue {
+	q := Queue{
```
Comment (stevenctl) [inline pkg/kube/controllers/queue.go:70]: nit: Is there a reason we don't want pointer receiver or return a pointer here?

File: pkg/kube/controllers/queue.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	return func(q *Queue) {
+		q.maxAttempts = n
+	}
+}
+
+// WithReconciler defines the to handle items on the queue
+func WithReconciler(f func(name types.NamespacedName) error) func(q *Queue) {
+	return func(q *Queue) {
+		q.work = func(key interface{}) error {
+			return f(key.(types.NamespacedName))
```
Comment (stevenctl) [inline pkg/kube/controllers/queue.go:63]: Is it better to type assert here and return an error? 
If there is no way to enqueue a `work` func that doesn't use `NamespacedName`, should we even export `Add(item interface{}`?

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:94]: It references `dc` so I don't think it can?

File: pkg/kube/controllers/queue.go (hunk 3 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:63]: I was going to have a function that allowed `interface{}` but turned out that nothing needed it (yet??). Likely its a bad idea to use anything else, so maybe we should just clean this up and not export Add

File: pkg/kube/controllers/queue.go (hunk 2 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:70]: I don't know. Is there a reason we should return a pointer? Seems most Queues do so maybe thats reason enough

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go (hunk 1 above)
Comment (stevenctl) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:94]: Misread, thought `Reconcile` was a struct member on first readthrough.

File: pkg/kube/controllers/queue.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+		Namespace: obj.GetNamespace(),
+		Name:      obj.GetName(),
+	})
+}
+
+// Run the queue. This is synchronous, so should typically be called in a goroutine.
+func (q Queue) Run(stop <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer q.queue.ShutDown()
+	log.Infof("starting %v", q.name)
```
Comment (Monkeyanator) [inline pkg/kube/controllers/queue.go:105]: nit: could store a logger on the struct with queue name prefix instead of prepending everywhere

File: pkg/revisions/default_watcher.go [hunk 5]
```diff
@@ -159,20 +100,19 @@ func (p *defaultWatcher) notifyHandlers() {
...
 	p.mu.Lock()
 	defer p.mu.Unlock()
 	p.defaultRevision = revision
 	p.notifyHandlers()
+	return nil
 }
 
-func filterUpdate(obj metav1.Object) bool {
-	return obj.GetName() != defaultTagWebhookName
+func filterUpdate(obj controllers.Object) bool {
```
Comment (Monkeyanator) [inline pkg/revisions/default_watcher.go:116]: very nit: counterintuitive naming, filterUpdate should be true when the object is not the default tag

File: pkg/revisions/default_watcher.go [hunk 6]
```diff
@@ -53,35 +49,29 @@ type defaultWatcher struct {
...
 func NewDefaultWatcher(client kube.Client, revision string) DefaultWatcher {
 	p := &defaultWatcher{
-		revision:    revision,
-		queue:       workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
-		initialSync: atomic.NewBool(false),
-		mu:          sync.RWMutex{},
+		revision: revision,
+		mu:       sync.RWMutex{},
 	}
+	p.queue = controllers.NewQueue(controllers.WithName("default revision"), controllers.WithReconciler(p.setDefault))
```
Comment (Monkeyanator) [inline pkg/revisions/default_watcher.go:62]: How do we detect default webhook deletion now? Before if we got a delete event, we'd call `setDefault` with `""`, now it seems like we'd set the default to the last active default?

File: pkg/revisions/default_watcher.go (hunk 5 above)
Comment (howardjohn) [reply pkg/revisions/default_watcher.go:116]: I would say `filter` is a fairly well known pattern in functional programing https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#filter-java.util.function.Predicate- 

But I will admit I confused myself when writing the code though so I mostly agree

File: pkg/revisions/default_watcher.go (hunk 6 above)
Comment (howardjohn) [reply pkg/revisions/default_watcher.go:62]:
```
	revision := ""
	wh, _, _ := p.webhookInformer.GetIndexer().GetByKey(key.Name)
	if wh != nil {
		revision = wh.(metav1.Object).GetLabels()[label.IoIstioRev.Name]
	}
```

if its delete `wh` is nil and therefor revision is `""`

File: pkg/kube/controllers/queue.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
+// Copyright Istio Authors
```
Comment (hzxuzhonghu) [inline pkg/kube/controllers/queue.go:1]: how about putting it under ` pkg/kube/queue`

File: pkg/kube/controllers/queue.go [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+// Queue defines an abstraction around Kubernetes' workqueue.
+// Items enqueued are deduplicated; this generally means relying on ordering of events in the queue is not feasible.
+type Queue struct {
+	queue       workqueue.RateLimitingInterface
+	initialSync *atomic.Bool
+	name        string
+	maxAttempts int
+	work        func(key interface{}) error
```
Comment (hzxuzhonghu) [inline pkg/kube/controllers/queue.go:35]: s/work/worker

File: pkg/kube/controllers/queue.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pkg/kube/controllers/queue.go:63]: why not the args consistent?

File: pkg/kube/controllers/queue.go [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+// NewQueue creates a new queue
+func NewQueue(options ...func(*Queue)) Queue {
+	q := Queue{
+		initialSync: atomic.NewBool(false),
+	}
+	for _, o := range options {
+		o(&q)
+	}
+	if q.name == "" {
+		q.name = "queue"
```
Comment (hzxuzhonghu) [inline pkg/kube/controllers/queue.go:77]: nit: It seems meaningless

File: pkg/kube/controllers/queue.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+// Run the queue. This is synchronous, so should typically be called in a goroutine.
+func (q Queue) Run(stop <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer q.queue.ShutDown()
+	log.Infof("starting %v", q.name)
+	q.Add(defaultSyncSignal)
+	done := make(chan struct{})
+	go func() {
+		// Process updates until we return false, which indicates the queue is terminated
```
Comment (hzxuzhonghu) [inline pkg/kube/controllers/queue.go:109]: defer utilruntime.HandleCrash() should be here

File: pkg/kube/controllers/queue.go [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+		for q.processNextItem() {
+		}
+		close(done)
```
Comment (hzxuzhonghu) [inline pkg/kube/controllers/queue.go:112]: do we have cases to return?

File: pkg/kube/controllers/queue.go [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+	case <-done:
+	}
+	log.Infof("stopped %v", q.name)
+}
+
+// syncSignal defines a dummy signal that is enqueued when .Run() is called. This allows us to detect
+// when we have processed all items added to the queue prior to Run().
+type syncSignal struct{}
+
+var defaultSyncSignal = syncSignal{}
```
Comment (hzxuzhonghu) [inline pkg/kube/controllers/queue.go:125]: Still confused with the sync stuff just seeing the queue itself

File: pkg/kube/controllers/queue.go (hunk 9 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:77]: it's for logging so we can distinguish queues

File: pkg/kube/controllers/queue.go (hunk 12 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:125]: What is the confusion? I can help update the comments but wasn't sure. This is taking the sync logic we already have in a lot of queues

File: pkg/kube/controllers/queue.go (hunk 11 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:112]: I think we need to in case the queue is closed. TBH this should probably never happen but we cannot

File: pkg/kube/controllers/queue.go (hunk 10 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:109]: Hm, I am not sure. Controllers in kubernetes does not do this; they look like the code here

File: pkg/kube/controllers/queue.go (hunk 7 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:1]: A bit worried about confusion with `pkg/queue`... maybe we can move it after we move everything over?

File: pkg/kube/controllers/queue.go (hunk 3 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:63]: I don't know what you mean by args consistent?

Comment (hzxuzhonghu) [reply pkg/kube/controllers/queue.go:66]: interface{} vs types.NamespacedName

File: pkg/kube/controllers/queue.go (hunk 7 above)
Comment (hzxuzhonghu) [reply pkg/kube/controllers/queue.go:1]: sgtm

File: pkg/kube/controllers/queue.go (hunk 10 above)
Comment (hzxuzhonghu) [reply pkg/kube/controllers/queue.go:113]: If keep defer outside the go routine, it could not capture the crash users passed in and executed in this routine

File: pkg/kube/controllers/queue.go (hunk 12 above)
Comment (hzxuzhonghu) [reply pkg/kube/controllers/queue.go:129]: It means to process all the buffering items enqueued before serving ready i think.

File: pilot/pkg/bootstrap/sidecarinjector.go [hunk 13]
```diff
@@ -99,7 +99,7 @@ func (s *Server) initSidecarInjector(args *PilotArgs) (*inject.Webhook, error) {
 				return nil
 			}
 
-			patcher.Run(stop)
+			go patcher.Run(stop)
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/sidecarinjector.go:102]: unrelated?

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go [hunk 14]
```diff
@@ -239,7 +239,7 @@ func (m *Multicluster) ClusterAdded(cluster *multicluster.Cluster, clusterStopCh
 			if err != nil {
 				log.Errorf("could not initialize webhook cert patcher: %v", err)
 			} else {
-				patcher.Run(clusterStopCh)
+				go patcher.Run(clusterStopCh)
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/multicluster.go:242]: ditto

File: pkg/kube/controllers/queue.go (hunk 3 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:66]: NamespacedName is the external interface. Internally we pass sync signal as well so it's generic

File: pkg/kube/controllers/queue.go (hunk 10 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:113]: Maybe we don't need it then.. every controller I looked at does it this way

File: pkg/kube/controllers/queue.go (hunk 12 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:129]: Yep

File: pkg/revisions/default_watcher.go [hunk 15]
```diff
@@ -53,35 +49,30 @@ type defaultWatcher struct {
...
 func NewDefaultWatcher(client kube.Client, revision string) DefaultWatcher {
 	p := &defaultWatcher{
-		revision:    revision,
-		queue:       workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
-		initialSync: atomic.NewBool(false),
-		mu:          sync.RWMutex{},
+		revision: revision,
+		mu:       sync.RWMutex{},
 	}
+	log.Errorf("howardjohn: default watch 1")
```
Comment (hzxuzhonghu) [inline pkg/revisions/default_watcher.go:62]: REMOVE

File: pilot/pkg/bootstrap/sidecarinjector.go (hunk 13 above)
Comment (howardjohn) [reply pilot/pkg/bootstrap/sidecarinjector.go:102]: Just keeping consistent pattern for controllers to have Run block

**Review by stevenctl — APPROVED** [review]

**Review by hzxuzhonghu — APPROVED** [review]

File: pkg/kube/controllers/queue.go [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+		Namespace: obj.GetNamespace(),
+		Name:      obj.GetName(),
+	})
+}
+
+// Run the queue. This is synchronous, so should typically be called in a goroutine.
+func (q Queue) Run(stop <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer q.queue.ShutDown()
+	q.log.Infof("starting")
```
Comment (nmittler) [inline pkg/kube/controllers/queue.go:109]: This doesn't seem like a very useful message. Should we be able to name our queues? Or maybe log this from the call sites instead?

File: pkg/kube/controllers/queue.go [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+func (q Queue) AddObject(obj Object) {
+	q.queue.Add(types.NamespacedName{
+		Namespace: obj.GetNamespace(),
+		Name:      obj.GetName(),
+	})
+}
+
+// Run the queue. This is synchronous, so should typically be called in a goroutine.
+func (q Queue) Run(stop <-chan struct{}) {
+	defer utilruntime.HandleCrash()
```
Comment (nmittler) [inline pkg/kube/controllers/queue.go:107]: What does this do?

File: pkg/kube/controllers/queue.go [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+		// Process updates until we return false, which indicates the queue is terminated
+		for q.processNextItem() {
+		}
+		close(done)
+	}()
+	select {
+	case <-stop:
+	case <-done:
+	}
+	q.log.Infof("stopped")
```
Comment (nmittler) [inline pkg/kube/controllers/queue.go:122]: Same comment as above

File: pkg/kube/controllers/queue.go (hunk 16 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:109]: `q.log` already embeds about the queues name and adds it to all messages

File: pkg/kube/controllers/queue.go (hunk 17 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:107]: I think for our case effectively nothing. I will remove it

File: pkg/kube/controllers/queue.go [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+func (q Queue) processNextItem() bool {
+	// Wait until there is a new item in the working queue
+	key, quit := q.queue.Get()
+	if quit {
+		// We are done, signal to exit the queue
+		return false
+	}
+
+	// We got the sync signal. This is not a real event, so we exit early after signaling we are synced
+	if key == defaultSyncSignal {
```
Comment (nmittler) [inline pkg/kube/controllers/queue.go:149]: So theoretically, any empty struct would evaluate to true here, correct? I guess it wouldn't make sense for a user to add an empty struct, but doing so could lead to a false positive for syncing?

File: pkg/kube/controllers/queue.go [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+		q.log.Debugf("synced")
```
Comment (nmittler) [inline pkg/kube/controllers/queue.go:150]: Same comment ... starting to feel like we should just provide a name to all of the queues?

File: pkg/kube/controllers/queue.go [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+	}
+
+	q.log.Debugf("handling update: %v", key)
+
+	// 'Done marks item as done processing' - should be called at the end of all processing
+	defer q.queue.Done(key)
+
+	err := q.workFn(key)
+	if err != nil {
+		if q.queue.NumRequeues(key) < q.maxAttempts {
```
Comment (nmittler) [inline pkg/kube/controllers/queue.go:162]: Do we know how many of our events are really retriable? If one attempt fails, will it ever succeed later?

File: pkg/kube/controllers/queue.go (hunk 19 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:149]: I think it should be impossible for any user outside of the `controllers` package. The queue only exposes interfaces to add Object or NamespacedName, so they cannot pass it in. And even if they did somehow, `syncSignal` is private as well so cannot be constructed by anyone else

File: pkg/kube/controllers/queue.go (hunk 16 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:109]: Also going to change it so name is required though

File: pkg/kube/controllers/queue.go (hunk 21 above)
Comment (howardjohn) [reply pkg/kube/controllers/queue.go:162]: Some of them will (api server writes). I think a safer default is to not retry, and then opt into retries for those that makes sense

**Review by nmittler — APPROVED** [review]
